lib/src/assemble/deployment_unit.rs

```diff
@@ -0,0 +1,253 @@
+                duplicates: diff
+                    .common
+                    .into_iter()
+                    .map(|(k, v)| (k.clone(), v.cloned()))
+                    .collect(),
+            });
+        }
+
+        Ok(NewDeploymentUnits { base: &mut self.deployment_units, units })
+    }
+
+    /// Convenience method for `self.start_add_deployment_unit(..).insert()` for a
+    /// single deployment unit.
+    pub fn insert(
+        &mut self,
+        data: DeploymentUnitData,
+    ) -> Result<(), DuplicateDeploymentUnitError> {
+        self.start_insert(data)?.commit();
+        Ok(())
+    }
+
+    pub fn finish_map(self) -> DeploymentUnitMap {
+        self.deployment_units
+    }
+
+    pub fn finish_units(self) -> ArtifactDeploymentUnits {
+        ArtifactDeploymentUnits::Composite {
+            deployment_units: self.deployment_units,
+        }
+    }
+}
+
+/// Information about new deployment units for a [`DeploymentUnitMapBuilder`].
+///
+/// This serves as a way to make new deployment units ready to be bulk-inserted
+/// into the builder.
+#[must_use = "NewDeploymentUnits must be committed into the builder"]
+pub struct NewDeploymentUnits<'a> {
+    base: &'a mut DeploymentUnitMap,
+    units: DeploymentUnitMap,
+}
+
+impl<'a> NewDeploymentUnits<'a> {
+    fn new_single(
+        base: &'a mut DeploymentUnitMap,
+        data: DeploymentUnitData,
+    ) -> Self {
+        let mut units = DeploymentUnitMap::new();
+        units.insert(data.hash_id(), data);
+        Self { base, units }
+    }
+
+    /// Inserts the deployment unit data into the builder.
+    pub fn commit(self) {
+        self.base.extend(self.units);
+    }
+}
+
+#[derive(Clone, Debug, Error)]
+pub struct DuplicateDeploymentUnitError {
+    /// The scope within which duplicates were found.
+    pub scope: DeploymentUnitScope,
+
+    /// Duplicates found while inserting deployment unit data.
+    ///
+    /// For `Leaf<DeploymentUnitData>`, `before` is the existing data and
+    /// `after` is the new data.
+    pub duplicates: BTreeMap<ArtifactHashId, Leaf<DeploymentUnitData>>,
+}
+
+impl DuplicateDeploymentUnitError {
+    // Note: existing and new should have the same hash_id.
+    fn new_single(
+        scope: DeploymentUnitScope,
+        existing: DeploymentUnitData,
+        new: DeploymentUnitData,
+    ) -> Self {
+        let mut duplicates = BTreeMap::new();
+        let hash_id = existing.hash_id();
+        duplicates.insert(hash_id, Leaf { before: existing, after: new });
+        Self { scope, duplicates }
+    }
+}
+
+impl fmt::Display for DuplicateDeploymentUnitError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // For a single deployment unit, we can simply display the artifact kind and hash.
+        if self.duplicates.len() == 1 {
+            let (hash_id, data) = self.duplicates.first_key_value().unwrap();
+            // XXX: should hash_id.kind/hash be a `Display` impl?
```

Comment:
I don't understand this comment - isn't this using their Display impls?

Reply:
Sorry, what I meant was whether we should add a `Display` impl to `ArtifactHashId`, or maybe have a `.display()` or `.display_human()` method on it.

Reply:
Updated the comment. It's pretty small, but would still love your thoughts on whether it makes sense to have a displayer hanging off of the `ArtifactHashId`.

Reply:
Got it. I think I'd vote for a `Display` impl; that would also be handy if these end up getting slog'd somewhere.

Reply:
Added it, and changed the format to be `kind/hash`, e.g. `zone/584bcfabf852d22ca322ca92076836cc8c9aa0489bc4132fef345271f2a00daf`.